src/main.rs

```diff
@@ -1,46 +1,156 @@
 use datafusion::functions_aggregate::min_max::max_udaf;
-use datafusion::logical_expr::ScalarUDF;
+use datafusion::logical_expr::{Signature, TypeSignature, Volatility};
 use datafusion::prelude::SessionContext;
+use std::any::Any;
 use std::sync::Arc;
 
 mod llm;
 
 #[tokio::main]
 async fn main() -> Result<()> {
     env_logger::init();
+
     let mut state = SessionStateBuilder::default()
-        .with_analyzer_rules(vec![Arc::new(LLMFunctionAnalyzeRule {})])
-        .with_optimizer_rules(vec![])
-        .with_query_planner(Arc::new(LLMQueryPlanner {}))
-        .with_physical_optimizer_rules(vec![])
+        .with_physical_optimizer_rule(Arc::new(AsyncFuncRule {}))
         .build();
-    let udf = ByPassScalarUDF::new("llm_bool", datafusion::arrow::datatypes::DataType::Boolean);
-    state.register_udf(Arc::new(ScalarUDF::new_from_impl(udf)))?;
+
+    let llm_bool = LLMBool::new();
+    let udf = AsyncScalarUDF::new(Arc::new(llm_bool));
+    state.register_udf(udf.into_scalar_udf())?;
     state.register_udaf(max_udaf())?;
     let ctx = SessionContext::new_with_state(state);
     ctx.sql("create table t1 (c1 int, c2 int, c3 int)")
         .await?
         .show()
         .await?;
-    ctx.sql("insert into t1 values (1, 2, 3), (1, 2, 3), (1, 2, 3)")
+    ctx.sql("insert into t1 values (1, 2, 3), (11, 2, 3), (1, 2, 3)")
         .await?
         .show()
         .await?;
-    ctx.sql("insert into t1 values (1, 2, 3), (1, 2, 3), (1, 2, 3)")
+    ctx.sql("insert into t1 values (1, 2, 3), (1, 2, 3), (21, 2, 3)")
         .await?
         .show()
         .await?;
-    ctx.sql("insert into t1 values (1, 2, 3), (1, 2, 3), (1, 2, 3)")
+    ctx.sql("insert into t1 values (31, 2, 3), (1, 2, 3), (1, 2, 3)")
         .await?
         .show()
         .await?;
 
     ctx.sql("explain select llm_bool('If all of them are Aisa countries: {}, {}, {}', t1.c1, t1.c2, t1.c3) from t1")
         .await?.show().await?;
+
+    ctx.sql("select llm_bool('If all of them are Aisa countries: {}, {}, {}', t1.c1, t1.c2, t1.c3) from t1")
+        .await?.show().await?;
+
     Ok(())
 }
+
+/// This is a simple example of a UDF that takes a string, invokes a (remote) LLM function
+/// and returns a boolean
+#[derive(Debug)]
+struct LLMBool {
+    signature: Signature,
+}
+
+impl LLMBool {
+    fn new() -> Self {
+        Self {
+            signature: Signature::one_of(
+                vec![TypeSignature::VariadicAny, TypeSignature::Nullary],
+                Volatility::Volatile,
+            ),
+        }
+    }
+}
+
+#[async_trait]
+impl AsyncScalarUDFImpl for LLMBool {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "llm_bool"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
+        Ok(DataType::Boolean)
+    }
+
+    async fn invoke_async(&self, args: &RecordBatch) -> Result<ArrayRef> {
```

Comment:
Here is the function that is invoked (it is `async`) and should be able to do any network and other calls